### ISSUE TITLE
Custom autograd sums output along batch dimension instead of failing

### ISSUE THREAD

**Issue by samdow** [pytorch/pytorch#65244]
I would expect the following code to error because the grad output (of shape [4,3]) doesn't match the input (of shape [3]). Instead, it sums along the first dimension, resulting, in this case, in an output of [4, 4, 4]

```python
class TestFunction(torch.autograd.Function):
  @staticmethod
  def forward(ctx, *unexpanded_args):
    return torch.tensor(1.)
  
  @staticmethod
  def backward(ctx, grad_output):
    return (torch.ones([4,3]), )

y = torch.randn(3, requires_grad=True)
x = TestFunction.apply(y)
x.backward()
y.grad # [4, 4, 4]
```

Not sure if there's a historical reason for summing along a batch dimension, but we ran into it while debugging some work for per sample gradients, so figured we would bring it up

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7

**Comment by albanD** [comment]
Hi,

I'm afraid this is expected because of broadcasting. The forward is allowed to do broadcasting without needing the formula to know about that.

Using `gradcheck` should catch such a problem though no?

**Comment by samdow** [comment]
Might be missing something but won't `gradcheck` assume that the gradients are in the `.grad` field? For what we have now, per sample gradients are in a different `.sample_grad`. Also from a quick look, I think there might be some other issues because the current ExpandedWeights mechanism is not really a tensor (so things like dtype don't exist).

From this, we are adding a test that the sum of the per sample gradients matches a batched grad call--maybe that's what you meant?

**Comment by soulitzer** [comment]
gradcheck uses `.grad` instead of `.backward` so it won't actually populate the `.grad` field.

**Comment by albanD** [comment]
@samdow I'm afraid the backward pass is "hard wired" to compute gradients and cannot compute per-sample gradients. In particular, the gradient shape is enforced to be the same as the forward Tensor.

You can see examples of repo that do similar things like https://github.com/cybertronai/autograd-hacks or https://backpack.pt/

Does that help?
Maybe you can share more about what you want to do?

**Comment by samdow** [comment]
Seems like this is the expected behavior so closing the issue
